src/human_encoding/README.md

```diff
@@ -0,0 +1,148 @@
+* EXCEPT for the following reserved symbols, which may not be used:
+    * `_`;
+    * `assertl`, `assertr`, `case`, `comp`, `disconnect`, `drop`, `fail`, `iden`, `injl`, `injr`, `pair`, `take`, `unit`, `witness`; and
+    * anything beginning with `jet_`.
+
+and EXPRESSION is
+
+* a NAME;
+* `unit`, `iden`, or `witness`;
+* `injl`, `injr`, `take`, `drop`, or `jet_word` followed by another EXPRESSION;
+* `case`, `comp`, or `pair` followed by two EXPRESSIONs;
+* `assertl` followed by an EXPRESSION, a literal `#`, and another EXPRESSION;
+* `assertr` followed by a literal `#` and two EXPRESSIONs;
+* a jet, which begins with `jet_` and must belong to the list of jets (FIXME define this list);
+* `jet_word` followed by a VALUE (defined below);
+* `fail` followed by an ENTROPY (defined below); or
+* `(` followed by another EXPRESSION followed by `)`.
+
+Note that while we allow parenthesis to help group parts of expressions for human
+understanding, they are never needed for disambiguation and are essentially
+ignored by the parser.
+
+A VALUE is one of
+* the literal `_`, which is interpreted as the empty value;
+* a binary literal `0b[01]+` which is interpreted as a sequence of bits; or
+* a hex literal `0x[0-9a-f]+` which is interpreted as a sequence of 4-bit nybbles
+
+An ENTROPY is a VALUE whose size is between 128 and 512 bits inclusive. Internally
+it is 0-padded out to 512 bits.
+
+Finally, TYPE is
+
+* a literal `_`, indicating no type bound;
+* a NAME;
+* a literal `1`, indicating the unit type;
+* a literal `2`, indicating the bit type;
+* `2^n`, where `n` is any power of two, in decimal with no spaces or punctuation;
+* `(` followed by another TYPE followed by `)`;
+* another TYPE, followed by `+`, followed by another TYPE; or
+* another TYPE, followed by `*`, followed by another TYPE.
+
+Here `*` has higher precedence than `+`, and both `+` and `*` are left-associative.
+
+## Semantics: Definitions
+
+As above, Simplicity expressions are a series of **definitions** of the form
+
+    NAME := EXPRESSION :: TYPE -> TYPE
+
+We refer to the `NAME` part as the **name**, `EXPRESSION` as the **expression**,
+and `TYPE -> TYPE` as the **type ascription**. If such a definition appears anywhere
+in a Simplicity encoding, then whenever that name appears in the expression of
+any other definition, its expression is substituted in place of it.
+
+For example, in the encoding
+
+    node := unit
+    main := comp node node
+
+the name `node` is substituted by `unit` both places that it appears. We can see
+that starting from the name `main`, by recursive substitution, we obtain a single
+Simplicity expression.
+
+The following restrictions apply:
+
+* The name `main` is special and must occur once in every encoding; recursive substitution starts from this symbol.
+* No cycles are allowed; if a name occurs anywhere in the expansion of its expression, this is an error.
+* Every name must be reachable from `main`. FIXME this is currently not enforced.
```

Comment:
Need to clarify that names can be defined multiple times, but exactly one must have an expression. (The others, if any, can add type bounds.)